File: lc_soln/0102_binary_tree_level_order_traversal/solution.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<vector<int>> levelOrder(TreeNode* root) {
        std::vector<std::vector<int>> result;
        if (root == nullptr) {
            return result;
        }
        std::deque<TreeNode*> q;
        q.push_back(root);
        while (!q.empty()) {
            int n = q.size();
            std::vector<int> row;
            while (n-- > 0) {
                TreeNode* node = q.front();
                q.pop_front();
                row.push_back(node->val);
                if (node->left != nullptr) {
                    q.push_back(node->left);
                }
                if (node->right != nullptr) {
                    q.push_back(node->right);
                }
            }
            result.push_back(row);
        }
        return result;
    }
};

int main() {
    return 0;
}

File: lc_soln/0102_binary_tree_level_order_traversal/solution.rs
struct Solution;

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//     pub val: i32,
//     pub left: Option<Rc<RefCell<TreeNode>>>,
//     pub right: Option<Rc<RefCell<TreeNode>>>,
// }
//
// impl TreeNode {
//     #[inline]
//     pub fn new(val: i32) -> Self {
//         TreeNode {
//             val,
//             left: None,
//             right: None,
//         }
//     }
// }

use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
        let mut result = Vec::new();
        let mut q = VecDeque::new();
        if let Some(node) = root {
            q.push_back(node);
        }

        while !q.is_empty() {
            let n = q.len();
            let mut row = Vec::new();
            for _ in 0..n {
                let node = q.pop_front().unwrap();
                let node = node.borrow();
                row.push(node.val);
                if let Some(ref left) = node.left {
                    q.push_back(Rc::clone(left));
                }
                if let Some(ref right) = node.right {
                    q.push_back(Rc::clone(right));
                }
            }
            result.push(row);
        }
        result
    }
}

fn main() {}
